include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+        return_in6_addr->u.Byte[12] = ipv4_inaddr->S_un.S_un_b.s_b1;
+        return_in6_addr->u.Byte[13] = ipv4_inaddr->S_un.S_un_b.s_b2;
+        return_in6_addr->u.Byte[14] = ipv4_inaddr->S_un.S_un_b.s_b3;
+        return_in6_addr->u.Byte[15] = ipv4_inaddr->S_un.S_un_b.s_b4;
+
+        return return_ipv6_address;
+    }
+
+    // non-member swap
+    inline void swap(::wil::network::socket_address& lhs, ::wil::network::socket_address& rhs) WI_NOEXCEPT
+    {
+        lhs.swap(rhs);
+    }
+
+    // class addr_info encapsulates the ADDRINFO-related structures returned from the socket functions
+    // getaddrinfo, GetAddrInfoW, GetAddrInfoWEx
+    // iterator semantics are supported to safely access these addresses
+    // template T supports pointers to the 3 address structures: ADDRINFOA*, ADDRINFOW*, ADDRINFOEXW*
+    template <typename T>
+    class addr_info_iterator_t
+    {
+    public:
+        // defining iterator_traits allows STL <algorithm> functions to be used with this iterator class.
+        // Notice this is a forward_iterator
+        // - does not support random-access (e.g. vector::iterator)
+        // - does not support bidirectional access (e.g. list::iterator)
+#if WIL_USE_STL || defined(WIL_DOXYGEN)
+        using iterator_category = ::std::forward_iterator_tag;
+#endif
+        using value_type = ::wil::network::socket_address;
+        using difference_type = size_t;
+        using distance_type = size_t;
+        using pointer = ::wil::network::socket_address*;
+        using reference = ::wil::network::socket_address&;
+
+        constexpr addr_info_iterator_t() WI_NOEXCEPT = default;
+
+        explicit addr_info_iterator_t(const T* addrinfo) WI_NOEXCEPT :
+            // must const cast so we can re-use this pointer as we walk the list
+            m_addrinfo_ptr(const_cast<T*>(addrinfo))
+        {
+            if (m_addrinfo_ptr)
+            {
+                m_socket_address.set_sockaddr(m_addrinfo_ptr->ai_addr, m_addrinfo_ptr->ai_addrlen);
+            }
+        }
+
+        ~addr_info_iterator_t() WI_NOEXCEPT = default;
+        addr_info_iterator_t(const addr_info_iterator_t&) WI_NOEXCEPT = default;
+        addr_info_iterator_t& operator=(const addr_info_iterator_t&) WI_NOEXCEPT = default;
+        addr_info_iterator_t(addr_info_iterator_t&&) WI_NOEXCEPT = default;
+        addr_info_iterator_t& operator=(addr_info_iterator_t&&) WI_NOEXCEPT = default;
+
+        const ::wil::network::socket_address& operator*() const WI_NOEXCEPT
+        {
+            return m_socket_address;
+        }
+
+        const ::wil::network::socket_address& operator*() WI_NOEXCEPT
+        {
+            return m_socket_address;
+        }
+
+        const ::wil::network::socket_address* operator->() const WI_NOEXCEPT
+        {
+            return &m_socket_address;
+        }
+
+        const ::wil::network::socket_address* operator->() WI_NOEXCEPT
+        {
+            return &m_socket_address;
+        }

```

Comment:
Having both these functions defined is unnecessary given 
```suggestion
        const ::wil::network::socket_address* operator->() const WI_NOEXCEPT
        {
            return &m_socket_address;
        }
```
Non-const function is unnecessary given that the return type of both is identical

Reply:
While you're at it, update the return types of this and `operator*` to return `pointer` and `reference` respectively

Reply:
Except the socket_address that the caller is calling operator-> might not be const. So I needed a non-const method so it can be called on both.

Reply:
I assume you mean `addr_info_iterator_t`, not `socket_address`? If that's the case, that's not how member functions work. You can call a `const` function on a non-`const` object (or pointer/reference to a maybe `const` object).

Reply:
ugh. yeah. sorry, I was thinking backwards.